pkg/cmd/cli/cmd/volume.go

```diff
@@ -0,0 +1,596 @@
+	volumeLong = `Update volumes on a pod
+
+This command can add, update, list or remove volumes from containers in pods or
+any object that has a pod template (replication controllers or deployment configurations).
+You can specify a single object or multiple, and alter volumes on all containers or
+just those that match a wildcard.`
+
+	volumeExample = `  // Add new volume of type 'emptyDir' for deployment config 'registry' and mount under /opt inside the containers
+  // The volume name is auto generated
+  $ %[1]s volume dc/registry --add --mount-path=/opt
+
+  // Add new volume 'v1' with secret 'magic' for pod 'p1'
+  $ %[1]s volume pod/p1 --add --name=v1 -m /etc --type=secret --secret-name=magic
+
+  // Add new volume to pod 'p1' based on gitRepo (or other volume sources not supported by --type)
+  $ %[1]s volume pod/p1 --add -m /repo --source=<json-string>
+
+  // Change rc 'r1' to use persistent volume instead of emptyDir (vol name: v1)
+  $ %[1]s volume rc/r1 --add --name=v1 -t persistentVolumeClaim --claim-name=pvc1 --confirm
+
+  // Change pod 'p1' mount point to /data for volume v1
+  $ %[1]s volume pod p1 --add --name=v1 -m /data --confirm
+
+  // Remove all volumes for pod 'p1'
+  $ %[1]s volume pod/p1 --remove --confirm
+
+  // Remove volume 'v1' from deployment config 'registry'  
+  $ %[1]s volume dc/registry --remove --name=v1
+
+  // Unmount volume v1 from container c1 on pod p1 and remove the volume v1 if it is not referenced by any containers on pod p1
+  $ %[1]s volume pod/p1 --remove --name=v1 --containers=c1
+
+  // List volumes defined on replication controller 'r1'
+  $ %[1]s volume rc r1 --list
+
+  // List volumes defined on all pods
+  $ %[1]s volume pods --all --list
+
+  // Output json object with volume info for pod 'p1' but don't alter the object on server
+  $ %[1]s volume pod/p1 --add --name=v1 --mount=/opt -o json`
+)
+
+type VolumeOptions struct {
+	DefaultNamespace       string
+	Writer                 io.Writer
+	Mapper                 meta.RESTMapper
+	Typer                  runtime.ObjectTyper
+	RESTClientFactory      func(mapping *meta.RESTMapping) (resource.RESTClient, error)
+	UpdatePodSpecForObject func(obj runtime.Object, fn func(*kapi.PodSpec) error) (bool, error)
+
+	// Resource selection
+	Selector string
+	All      bool
+
+	// Operations
+	Add    bool
+	Remove bool
+	List   bool
+
+	// Common optional params
+	Name          string
+	Containers    string
+	Confirm       bool
+	Output        string
+	OutputVersion string
+
+	// Add op params
+	AddOpts *AddVolumeOptions
+}
+
+type AddVolumeOptions struct {
+	Type       string
+	MountPath  string
+	Path       string
+	SecretName string
+	ClaimName  string
+	Source     string
+}
+
+var typeChanged bool
+
+func NewCmdVolume(fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	addOpts := &AddVolumeOptions{}
+	opts := &VolumeOptions{AddOpts: addOpts}
+	cmd := &cobra.Command{
+		Use:     "volume RESOURCE/NAME --add|--remove|--list",
```

Comment:
I don't consider three too many.  One of those is the default, so it's really 2.

----- Original Message -----
> > +type AddVolumeOptions struct {
> > +	Type       string
> > +	MountPath  string
> > +	Path       string
> > +	SecretName string
> > +	ClaimName  string
> > +	Source     string
> > +}
> > +
> > +var typeChanged bool
> > +
> > +func NewCmdVolume(fullName string, f *clientcmd.Factory, out io.Writer)
> > *cobra.Command {
> > +	addOpts := &AddVolumeOptions{}
> > +	opts := &VolumeOptions{AddOpts: addOpts}
> > +	cmd := &cobra.Command{
> > +		Use:     "volume RESOURCE/NAME --add|--remove|--list",
> 
> We discussed this on a scrum the other day. My preference was subs, Clayton
> said it would depend on how many common flags there are. I think looking at
> the examples, subs makes more sense.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/3023/files#r32127842